rust/main/chains/hyperlane-starknet/src/validator_announce.rs

```diff
@@ -0,0 +1,225 @@
+
+use crate::contracts::validator_announce::ValidatorAnnounce as StarknetValidatorAnnounceInternal;
+use crate::error::HyperlaneStarknetError;
+use crate::{
+    build_single_owner_account, get_transaction_receipt, string_to_cairo_long_string,
+    to_strk_message_bytes, ConnectionConf, Signer, StarknetProvider,
+};
+use cainome::cairo_serde::EthAddress;
+
+impl<A> std::fmt::Display for StarknetValidatorAnnounceInternal<A>
+where
+    A: starknet::accounts::ConnectedAccount + Sync + std::fmt::Debug,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "{self:?}")
+    }
+}
+
+/// A reference to a ValidatorAnnounce contract on some Starknet chain
+#[derive(Debug)]
+#[allow(unused)]
+pub struct StarknetValidatorAnnounce {
+    contract: Arc<StarknetValidatorAnnounceInternal<SingleOwnerAccount<AnyProvider, LocalWallet>>>,
+    provider: StarknetProvider,
+    conn: ConnectionConf,
+}
+
+impl StarknetValidatorAnnounce {
+    /// Create a reference to a ValidatorAnnounce at a specific Starknet address on some
+    /// chain
+    pub fn new(
+        conn: &ConnectionConf,
+        locator: &ContractLocator,
+        signer: Signer,
+    ) -> ChainResult<Self> {
+        let account = build_single_owner_account(
+            &conn.url,
+            signer.local_wallet(),
+            &signer.address,
+            false,
+            locator.domain.id(),
+        );
+
+        let contract = StarknetValidatorAnnounceInternal::new(
+            locator
+                .address
+                .try_into()
+                .map_err(HyperlaneStarknetError::BytesConversionError)?,
+            account,
+        );
+
+        Ok(Self {
+            contract: Arc::new(contract),
+            provider: StarknetProvider::new(locator.domain.clone(), conn),
+            conn: conn.clone(),
+        })
+    }
+
+    /// Returns a ContractCall that processes the provided message.
+    async fn announce_contract_call(
+        &self,
+        announcement: SignedType<Announcement>,
+    ) -> ChainResult<(
+        Execution<'_, SingleOwnerAccount<AnyProvider, LocalWallet>>,
+        FieldElement,
+    )> {
+        let validator = FieldElement::from_byte_slice_be(&announcement.value.validator.to_vec())
+            .map_err(Into::<HyperlaneStarknetError>::into)?;
+        let storage_location = string_to_cairo_long_string(&announcement.value.storage_location)
+            .map_err(Into::<HyperlaneStarknetError>::into)?;
+        let signature_bytes = announcement.signature.to_vec();
+        let signature = &to_strk_message_bytes(&signature_bytes);
+
+        let tx = self
+            .contract
+            .announce(&EthAddress(validator), &storage_location, signature);
+
+        let gas_estimate = tx
+            .estimate_fee()
+            .await
+            .map_err(|e| HyperlaneStarknetError::AccountError(e.to_string()))?
+            .overall_fee;
+
+        let max_cost = gas_estimate * FieldElement::TWO;
```

Comment:
assuming the estimate is doubled as a buffer for any errors? Does the caller get refunded any unused gas?

Reply:
yes it does! that's the usual good practice on starknet because gas estimate is not that precise